### ISSUE TITLE
The gpu memory increases rapidly when using executor reshape function.

### ISSUE THREAD

**Issue by ascust** [apache/incubator-mxnet#2232]
Hi, 
I tried to use "reshape" of executor to dynamically change the input size. Then I found the memory increases rapidly and becomes out of memory after some iterations. I use the following code to reproduce this problem:

```
import mxnet as mx
import random
input_data = mx.symbol.Variable(name="data")
conv1 = mx.symbol.Convolution(data=input_data, kernel=(11, 11), stride=(4, 4), num_filter=200)

e = conv1.simple_bind(ctx=mx.gpu(0),
                   grad_req="write",
                   data=(1, 3, 600, 600))
for i in range(1000):
    temp_shape = (1,3, random.randint(600, 700),random.randint(600, 700))
    e = e.reshape(allow_up_sizing=True, partial_shaping=True, data=temp_shape)
    e.forward(is_train=True)
```

My gpu has 4g memory. As far as I can tell, this cannot exceed the limit. The reality is I got out of memory error very soon. @tqchen @piiswrong

**Comment by piiswrong** [comment]
add e.outputs[0].wait_to_read() after forward

**Comment by piiswrong** [comment]
@tqchen We get this kind of issues every week. Maybe add a limit to how many forward can be pushed to engine?

**Comment by tqchen** [comment]
A better documentation might be a better way to address this, limiting queues in engine by blocking will bring extra complexity to it

**Comment by tqchen** [comment]
@ascust If you add e.output[0].copyto(mx.cpu()).asnumpy() which will be needed when you really copy data out, that will impose a synchronization as well so wait_to_read is not needed

**Comment by ascust** [comment]
@piiswrong @tqchen Thanks for your replies. i tried both your suggestions and the issue remains. I found that if I use another gpu with 6 GB memory, there is no out of memory because the maximum gpu memory usage is about 4.2 GB, which is sufficient for this card but not for my previous one. What is interesting is that why it takes so much memory to reshape if the new executor shares the memory with the old one? 
If I change the shape to a fixed one instead of random one like this:
```
temp_shape = (1,3, 700,700)
```
there is no problem, even without wait_to_read or copyto. The memory usage is only 180 MB. I guess there should be some extra memory usage during the allocation of the new shaped arguments, but 180 MB compared with 4.2 GB seems a little big difference. 
This only happens when I frequently change the input size. Any help would be really appreciated.

**Comment by tqchen** [comment]
try to avoid allow_up_sizing and always allocate the largest executor

**Comment by ascust** [comment]
The problem is I am trying to train a network with various input size. I do not know in advance what size the input has and it would be not a good choice to sort the data based on their size and do the training using this order. Would it be possible to change the input size arbitrarily with not that much extra memory cost? I mean would it be possible to recycle the some cach memory more efficiently?

**Comment by Redneckgirl96** [comment]
Stop sending me emails or I will report youSent from my Huawei Mobile-------- Original Message --------Subject: Re: [dmlc/mxnet] The gpu memory increases rapidly when using executor reshape function. (#2232)From: Tianqi Chen To: dmlc/mxnet CC: @ascust If you add e.output[0].copyto(mx.cpu()).asnumpy() which will be needed when you really copy data out, that will impose a synchronization as well so wait_to_read is not needed

**Comment by piiswrong** [comment]
you can bind with the maximum size that fit in the memory and you won't need allow_up_sizing

**Comment by ascust** [comment]
Alright, I tried this and it worked. This is clever. Thanks guys.

**Comment by ascust** [comment]
It seems if the network has more complex structures, it becomes not that stable. Here I use the residual structure from "example" folder. I changed number of filters for each group to increase the capcity of the network. and I also remove the alst fc and pooling layer to allow different input shapes. If I bind the network with say data = (1, 3, 1000, 1000), the memory usage is around 1533MB when the input size is fixed. If I bind the network with this maximum input size and reshape it using different smaller sizes, I will get out of memory soon. I have only 4GB gpu memory. The problem seems that the new executor does not really share the memory with the old one. 
I have pasted code here, please have a look. @piiswrong @tqchen 

```
import mxnet as mx


def conv_factory(data, num_filter, kernel, stride, pad, act_type='relu', conv_type=0):
    if conv_type == 0:
        conv = mx.symbol.Convolution(data=data, num_filter=num_filter, kernel=kernel, stride=stride, pad=pad)
        bn = mx.symbol.BatchNorm(data=conv)
        act = mx.symbol.Activation(data=bn, act_type=act_type)
        return act
    elif conv_type == 1:
        conv = mx.symbol.Convolution(data=data, num_filter=num_filter, kernel=kernel, stride=stride, pad=pad)
        bn = mx.symbol.BatchNorm(data=conv)
        return bn


def residual_factory(data, num_filter, dim_match):
    if dim_match == True:  # if dimension match
        identity_data = data
        conv1 = conv_factory(data=data, num_filter=num_filter, kernel=(3, 3), stride=(1, 1), pad=(1, 1),
                             act_type='relu', conv_type=0)

        conv2 = conv_factory(data=conv1, num_filter=num_filter, kernel=(3, 3), stride=(1, 1), pad=(1, 1), conv_type=1)
        new_data = identity_data + conv2
        act = mx.symbol.Activation(data=new_data, act_type='relu')
        return act
    else:
        conv1 = conv_factory(data=data, num_filter=num_filter, kernel=(3, 3), stride=(2, 2), pad=(1, 1),
                             act_type='relu', conv_type=0)
        conv2 = conv_factory(data=conv1, num_filter=num_filter, kernel=(3, 3), stride=(1, 1), pad=(1, 1), conv_type=1)

        # adopt project method in the paper when dimension increased
        project_data = conv_factory(data=data, num_filter=num_filter, kernel=(1, 1), stride=(2, 2), pad=(0, 0),
                                    conv_type=1)
        new_data = project_data + conv2
        act = mx.symbol.Activation(data=new_data, act_type='relu')
        return act


def residual_net(data, n):
    # fisrt 2n layers
    for i in range(n):
        data = residual_factory(data=data, num_filter=64, dim_match=True)

    # second 2n layers
    for i in range(n):
        if i == 0:
            data = residual_factory(data=data, num_filter=128, dim_match=False)
        else:
            data = residual_factory(data=data, num_filter=128, dim_match=True)

    # third 2n layers
    for i in range(n):
        if i == 0:
            data = residual_factory(data=data, num_filter=256, dim_match=False)
        else:
            data = residual_factory(data=data, num_filter=256, dim_match=True)

    return data


def get_symbol(num_classes=10):
    conv = conv_factory(data=mx.symbol.Variable(name='data'), num_filter=64, kernel=(3, 3), stride=(1, 1), pad=(1, 1),
                        act_type='relu', conv_type=0)
    n = 3  # set n = 3 means get a model with 3*6+2=20 layers, set n = 9 means 9*6+2=56 layers
    resnet = residual_net(conv, n)  #
    return resnet

net = get_symbol()

e = net.simple_bind(ctx=mx.gpu(0),
                    grad_req="null",
                    data=(1, 3, 1000, 1000))
import random
for i in range(1000):
    temp_shape = (1, 3, random.randint(600, 900), random.randint(600, 900))
    e_t = e.reshape(data=temp_shape)
    e_t.forward(is_train=False)
    output = e_t.outputs[0].copyto(mx.cpu()).asnumpy()
```

**Comment by sxjscience** [comment]
@ascust I've run the code locally and have not experienced the OOM problem. The memory usage is around 1400MB. I'm using Windows 10 + GTX 980 Ti + Cuda 8.0

**Comment by ascust** [comment]
Ok Thanks for your reply. I found the problem might be this as shown below. Please try the below code, where there is only one conv layer with big dilation and num_filter. Normally, if I fix the input size, it uses only 800 MB memory, but if I vary the input size, more than 4000 MB will be used. There seems a lot of memory that could be squeezed. @sxjscience 
Try this out:
```
net = mx.symbol.Convolution(data=mx.symbol.Variable(name='data'), kernel=(3,3),pad=(12,12), num_filter=150, dilate=(12,12), workspace=2048)

e = net.simple_bind(ctx=mx.gpu(0),
                    grad_req="null",
                    data=(1, 2048, 80, 80))
import random
for i in range(1000):
    temp_shape = (1, 2048, random.randint(50, 80), random.randint(50, 80))
    e_t = e.reshape(data=temp_shape)
    e_t.forward(is_train=False)
    output = e_t.outputs[0].copyto(mx.cpu()).asnumpy()
```

**Comment by ascust** [comment]
@piiswrong @tqchen any thoughts guys? is that because the previous memory did not get released instantly? Just out of curiosity, since the new executor would share the memory, how come this could happen?

**Comment by ascust** [comment]
![screen shot 2016-08-09 at 3 06 04 pm](https://cloud.githubusercontent.com/assets/12552820/17505798/cf5a4832-5e42-11e6-9837-c0690587c71f.png)
I have made some tests for different dilation size and I found if dilation!=1, this question seemed to be very obvious. The table below shows the gpu memory for different dilation size with or without reshape (means fixed input size or various input size). We can see that when dilation=1, everything is very normal, but when it is equal to 2, there is a big memory usage, for various input size. fixed size has a big increase as well but not that significant. Then with the dilation size increases, the memory usage seems to drop.
The code I used to make the test was just my previous reply. Is this normal? Is there supposed to be big memory usage when dilation is bigger than 1? @piiswrong @tqchen

**Comment by piiswrong** [comment]
When dilation > 1 we use mxnet's own conv layer instead of cudnn because it
doesn't support dilation.
Could you try turn off cudnn and see if dilation still has this effect?

**Comment by ascust** [comment]
@piiswrong Without using Cudnn, this effect still exists. For example, using dilation=12, fixed input size uses 600MB, while various input size has more than 4GB usage, since my 4GB gpu is out of memory. More importantly, this time, even with dilation=1, the system got out of memory when there is various input size since cudnn is disabled.

**Comment by piiswrong** [comment]
Are you using current master? https://github.com/dmlc/mxnet/pull/2927 should fix this problem

**Comment by ascust** [comment]
@piiswrong I just updated mxnet to the lastest one and I dont think the issues has been fixed. Nevertheless, it seems the issue has been mitigated but still obvious. For example, with cudnn, if I use dilate=12, with fixed input size=(1,2048,80,80), the memory usage is around 600MB. However if I used reshape to get a new shared-weights executor with a new input size(even though the new size is equal to the previous one), the memory usage increased to 1800MB (three times). I have tried using other dilate values, I observed at lease 2 times bigger of memory usage.
